File: apps/ios/Sources/Networking/SaaSAPIClient.swift
import Foundation
import os.log

private let rpcLogger = Logger(subsystem: "com.thinkfleet", category: "rpc")

actor SaaSAPIClient {
    let baseURL: URL
    private let session: URLSession
    private let sessionStore: SessionStore

    init(baseURL: URL, sessionStore: SessionStore, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.sessionStore = sessionStore
        self.session = session
    }

    // MARK: - oRPC Protocol Envelope
    // oRPC RPC protocol wraps requests as {"json": <input>} and responses as {"json": <output>}

    private struct RPCRequest<T: Encodable>: Encodable { let json: T }
    private struct RPCResponse<T: Decodable>: Decodable { let json: T }
    private struct EmptyInput: Encodable {}

    // MARK: - oRPC Calls

    func rpc<T: Decodable>(_ path: String, input: Encodable? = nil) async throws -> T {
        let rpcPath = path.replacingOccurrences(of: ".", with: "/")
        let url = baseURL.appendingPathComponent("/api/rpc/\(rpcPath)")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        applyAuth(to: &request)

        if let input {
            request.httpBody = try encodeRPCInput(input)
        } else {
            request.httpBody = try JSONEncoder().encode(RPCRequest(json: EmptyInput()))
        }

        let (data, response) = try await session.data(for: request)

        if let httpResponse = response as? HTTPURLResponse {
            let bodyPreview = String(data: data.prefix(500), encoding: .utf8) ?? "non-utf8"
            rpcLogger.info("RPC \(rpcPath) → \(httpResponse.statusCode): \(bodyPreview)")
        }

        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 401 {
            let refreshed = await refreshSession()
            if refreshed {
                applyAuth(to: &request)
                let (retryData, retryResponse) = try await session.data(for: request)
                guard let retryHttp = retryResponse as? HTTPURLResponse, retryHttp.statusCode == 200 else {
                    throw APIError.requestFailed(statusCode: (retryResponse as? HTTPURLResponse)?.statusCode ?? 0)
                }
                return try decodeRPCResponse(retryData)
            }
            // Don't clear session here — the token may still be valid (server propagation delay).
            // The caller (e.g. loadOrganizations) can retry, and signOut handles clearing.
            throw APIError.unauthorized
        }

        guard let httpResponse = response as? HTTPURLResponse, (200 ..< 300).contains(httpResponse.statusCode) else {
            throw APIError.requestFailed(statusCode: (response as? HTTPURLResponse)?.statusCode ?? 0)
        }

        return try decodeRPCResponse(data)
    }

    /// Wrap input in oRPC envelope: {"json": <input>}
    private func encodeRPCInput(_ input: Encodable) throws -> Data {
        let inputData = try JSONEncoder().encode(input)
        let inputJson = try JSONSerialization.jsonObject(with: inputData)
        let wrapped: [String: Any] = ["json": inputJson]
        return try JSONSerialization.data(withJSONObject: wrapped)
    }

    /// Unwrap oRPC response envelope: {"json": <data>} → <data>
    private func decodeRPCResponse<T: Decodable>(_ data: Data) throws -> T {
        return try JSONDecoder().decode(RPCResponse<T>.self, from: data).json
    }

    // MARK: - Generic REST

    func get<T: Decodable>(path: String) async throws -> T {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        applyAuth(to: &request)
        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(T.self, from: data)
    }

    // MARK: - Helpers

    /// Cookie name used by Better Auth in production (https:// base URL enables __Secure- prefix).
    private static let sessionCookieName = "__Secure-better-auth.session_token"

    private func applyAuth(to request: inout URLRequest) {
        if let token = sessionStore.sessionToken {
            let cookie = "\(Self.sessionCookieName)=\(token)"
            let urlPath = request.url?.lastPathComponent ?? "?"
            rpcLogger.info("applyAuth: Cookie=\(cookie.prefix(50))... to \(urlPath)")
            request.setValue(cookie, forHTTPHeaderField: "Cookie")
        } else {
            rpcLogger.warning("applyAuth: no session token!")
        }
        // Origin header required by Better Auth for CSRF validation on POST requests
        let origin = "\(baseURL.scheme ?? "https")://\(baseURL.host ?? "")"
        request.setValue(origin, forHTTPHeaderField: "Origin")
        if let orgId = sessionStore.currentOrganizationId {
            request.setValue(orgId, forHTTPHeaderField: "x-organization-id")
        }
    }

    private func refreshSession() async -> Bool {
        guard let token = sessionStore.sessionToken else { return false }
        var request = URLRequest(url: baseURL.appendingPathComponent("/api/auth/get-session"))
        request.setValue("\(Self.sessionCookieName)=\(token)", forHTTPHeaderField: "Cookie")
        request.setValue("\(baseURL.scheme ?? "https")://\(baseURL.host ?? "")", forHTTPHeaderField: "Origin")

        guard let (_, response) = try? await session.data(for: request),
              let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200
        else { return false }

        // If server sends a refreshed signed token in Set-Cookie, update the stored token
        if let newSignedToken = Self.extractSessionToken(from: httpResponse) {
            await MainActor.run { sessionStore.setSession(token: newSignedToken, user: sessionStore.currentUser!) }
        }
        return true
    }

    /// Extract the signed session token from the Set-Cookie header.
    private static func extractSessionToken(from response: HTTPURLResponse) -> String? {
        guard let setCookie = response.value(forHTTPHeaderField: "Set-Cookie") else { return nil }
        let pattern = /better-auth\.session_token=([^;,]+)/
        guard let match = setCookie.firstMatch(of: pattern) else { return nil }
        return String(match.1)
    }
}

enum APIError: LocalizedError {
    case unauthorized
    case requestFailed(statusCode: Int)
    case decodingFailed

    var errorDescription: String? {
        switch self {
        case .unauthorized: "Authentication required."
        case let .requestFailed(code): "Request failed with status \(code)."
        case .decodingFailed: "Failed to parse response."
        }
    }
}

File: apps/android/app/src/main/java/bot/molt/android/networking/SaaSApiClient.kt
package bot.molt.android.networking

import android.util.Log
import bot.molt.android.auth.SessionStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.KSerializer
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

private const val TAG = "SaaSApi"

class SaaSApiClient(
    private val baseUrl: String,
    private val sessionStore: SessionStore,
    private val client: OkHttpClient = OkHttpClient(),
) {
    private val json = Json { ignoreUnknownKeys = true }

    private fun rpcUrl(path: String): String {
        val slashPath = path.replace('.', '/')
        return "$baseUrl/api/rpc/$slashPath"
    }

    // oRPC RPC protocol wraps requests as {"json": <input>} and responses as {"json": <output>}

    suspend fun <I, O> rpc(
        path: String,
        input: I,
        inputSerializer: KSerializer<I>,
        outputSerializer: KSerializer<O>,
    ): O = withContext(Dispatchers.IO) {
        val inputBody = json.encodeToString(inputSerializer, input)
        val body = """{"json":$inputBody}"""
        val url = rpcUrl(path)
        Log.i(TAG, "rpc: POST $path token=${sessionStore.sessionToken.value?.take(8)}... org=${sessionStore.organizationId.value}")
        val request = buildRequest(url, body)

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: ""
        Log.i(TAG, "rpc: $path → ${response.code}: ${responseBody.take(200)}")

        if (response.code == 401) {
            Log.w(TAG, "rpc: $path got 401, attempting session refresh")
            val refreshed = refreshSession()
            if (refreshed) {
                val retryRequest = buildRequest(url, body)
                val retryResponse = client.newCall(retryRequest).execute()
                val retryBody = retryResponse.body?.string() ?: ""
                Log.i(TAG, "rpc: $path retry → ${retryResponse.code}: ${retryBody.take(200)}")
                if (!retryResponse.isSuccessful) throw ApiException(retryResponse.code, "Request failed")
                return@withContext decodeRpcResponse(retryBody, outputSerializer)
            }
            throw ApiException(401, "Unauthorized")
        }

        if (!response.isSuccessful) throw ApiException(response.code, "Request failed")
        decodeRpcResponse(responseBody, outputSerializer)
    }

    suspend fun <O> rpcNoInput(
        path: String,
        outputSerializer: KSerializer<O>,
    ): O = withContext(Dispatchers.IO) {
        val url = rpcUrl(path)
        Log.i(TAG, "rpcNoInput: POST $path token=${sessionStore.sessionToken.value?.take(8)}... org=${sessionStore.organizationId.value}")
        val request = buildRequest(url, """{"json":{}}""")

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: ""
        Log.i(TAG, "rpcNoInput: $path → ${response.code}: ${responseBody.take(200)}")

        if (response.code == 401) {
            Log.w(TAG, "rpcNoInput: $path got 401, attempting session refresh")
            val refreshed = refreshSession()
            if (refreshed) {
                val retryRequest = buildRequest(url, """{"json":{}}""")
                val retryResponse = client.newCall(retryRequest).execute()
                val retryBody = retryResponse.body?.string() ?: ""
                Log.i(TAG, "rpcNoInput: $path retry → ${retryResponse.code}: ${retryBody.take(200)}")
                if (!retryResponse.isSuccessful) throw ApiException(retryResponse.code, "Request failed")
                return@withContext decodeRpcResponse(retryBody, outputSerializer)
            }
            throw ApiException(401, "Unauthorized")
        }

        if (!response.isSuccessful) throw ApiException(response.code, "Request failed")
        decodeRpcResponse(responseBody, outputSerializer)
    }

    /** Unwrap oRPC response envelope: {"json": <data>} → <data> */
    private fun <O> decodeRpcResponse(responseStr: String, outputSerializer: KSerializer<O>): O {
        val envelope = json.parseToJsonElement(responseStr).jsonObject
        val jsonContent = envelope["json"] ?: throw ApiException(0, "Invalid oRPC response: missing json field")
        return json.decodeFromJsonElement(outputSerializer, jsonContent)
    }

    private fun buildRequest(url: String, body: String): Request {
        val builder = Request.Builder()
            .url(url)
            .post(body.toRequestBody(JSON_MEDIA))
            .header("Content-Type", "application/json")
            .header("Accept", "application/json")
            // Origin header required by Better Auth for CSRF validation on POST requests
            .header("Origin", baseUrl.trimEnd('/'))

        sessionStore.sessionToken.value?.let { token ->
            builder.header("Cookie", "$SESSION_COOKIE_NAME=$token")
        }
        sessionStore.organizationId.value?.let { orgId ->
            builder.header("x-organization-id", orgId)
        }

        return builder.build()
    }

    private suspend fun refreshSession(): Boolean = withContext(Dispatchers.IO) {
        val token = sessionStore.sessionToken.value ?: return@withContext false
        Log.i(TAG, "refreshSession: attempting with token=${token.take(8)}...")
        val request = Request.Builder()
            .url("$baseUrl/api/auth/get-session")
            .header("Cookie", "$SESSION_COOKIE_NAME=$token")
            .header("Origin", baseUrl.trimEnd('/'))
            .build()

        val response = try { client.newCall(request).execute() } catch (e: Exception) {
            Log.e(TAG, "refreshSession: network error: ${e.message}")
            return@withContext false
        }
        Log.i(TAG, "refreshSession: status=${response.code}")
        if (!response.isSuccessful) return@withContext false

        // If server sends a refreshed signed token in Set-Cookie, update the stored token
        val newSignedToken = extractSessionToken(response)
        if (newSignedToken != null) {
            Log.i(TAG, "refreshSession: got refreshed signed token=${newSignedToken.take(8)}...")
            sessionStore.currentUser.value?.let { user ->
                sessionStore.setSession(newSignedToken, user = user)
            }
        }
        true
    }

    /** Extract the signed session token from the Set-Cookie header. */
    private fun extractSessionToken(response: okhttp3.Response): String? {
        val setCookies = response.headers("Set-Cookie")
        for (cookie in setCookies) {
            val regex = Regex("""better-auth\.session_token=([^;,]+)""")
            val match = regex.find(cookie)
            if (match != null) return match.groupValues[1]
        }
        return null
    }

    companion object {
        /** Cookie name used by Better Auth in production (https:// base URL enables __Secure- prefix). */
        private const val SESSION_COOKIE_NAME = "__Secure-better-auth.session_token"
        private val JSON_MEDIA = "application/json".toMediaType()
    }
}

class ApiException(val statusCode: Int, message: String) : Exception(message)
